### DISCUSSION TITLE
Unclear documentation about Git extension's interaction with virtual workspace

### DISCUSSION THREAD

**Discussion by jjyyxx** [microsoft/vscode-discussions#1012] [Extension Development QnA]
vscode's [FileSystemProvider](https://code.visualstudio.com/api/references/vscode-api#FileSystemProvider) provides a powerful way  to serve custom (remote, database, etc.) content. In built-in Git extension's [package.json](https://github.com/microsoft/vscode/blob/main/extensions/git/package.json#L53), it sets `"virtualWorkspaces": true`, indicating the intention to work with virtual workspaces, but it seems that Git extension does not work well with a custom FileSystemProvider.

Take `git.init` command as an example, it relies `workspace.workspaceFolders[0].uri.fsPath`, which directly contradicts the [documentation guideline](https://code.visualstudio.com/api/extension-guides/virtual-workspaces#review-that-the-extension-code-is-ready-for-virtual-resources):
> Never assume that the URI scheme is `file`. `URI.fsPath` can only be used when the URI scheme is file.
>
This behavior would fail at best with `ENOENT`, and at worst, might coincidentally modify local files or do unintended things.

I know that getting Git, a file-based VCS, to work correctly with VFS requires lots of extra work, but I have several question seeking clarification
- If it is bound to fail, the extension (or failing commands) should be properly disabled (maybe conditionally);
- If it can/cannot support virtual workspaces, I think the documentation should clearly state this;
- Is there any idiomatic way to connect the built-in Git extension to a custom virtual workspace? For example, under several scenarios:
  1. In the best case, automatic transparent support (even if at a cost of inferior performance)
  2. The extension providing the VFS can somehow providing functionality needed by Git, e.g., backed by a remotely running Git similar to Remote development extensions
  3. The extension providing the VFS can manage a sync-ed local mirror, (in this case, it is of course easy to provide support by openning the local mirror directory, but I wish that) Git extension use the local mirror to provide VCS support to the VFS
  4. `RemoteSource` API looks interesting to me by its name, but lacks documentation

**Comment by isidorn** [comment]
What are you trying to achieve in your extension?

**Reply by jjyyxx** [reply]
The core functionality is a virtual workspace mapped to certain web service with data organized in tree structure (directory-like). It overall works, but I would like to integrate with the Git extension, for easy version control. The objective is simple:
- Ideally: seamlessly integrate with Git extension, even if this requires extra code on my side (e.g., sync a local copy, write an adapter, or maintain a remote git "proxy")
- At least: Clarify that Git extension won't work with virtual workspace, and have a way to disable the broken functionalities (e.g., initialize repo)

**Reply by isidorn** [reply]
@lszomoru might be able to help here

**Reply by abrilurena** [reply]
Hi, @lszomoru! i hope you're doing great. Could you please help @jjyyxx with this? thank you!

**Comment by gjsjohnmurray** [comment]
I'm guessing the OP would like the Git extension to be able to manage the virtual files as though they are local files. AFAIK this isn't possible because the extension leverages the Git installed on your workstation, which has no knowledge of any virtual filesystems implemented by VS Code extensions.

**Reply by gjsjohnmurray** [reply]
Maybe some day the Git extension might be able to leverage [isomorphic-git](https://isomorphic-git.org/en) whose [pluggable filesystem](https://isomorphic-git.org/docs/en/fs) architecture looks like it could in turn leverage the `FileSystemProvider` API in VS Code, perhaps via `vscode.workspace.fs`

See also https://github.com/microsoft/vscode/issues/55110#issuecomment-455483436

**Reply by jjyyxx** [reply]
Thanks for your clarification! [isomorphic-git](https://isomorphic-git.org/en) looks promising, but also has a long way to go. Then I'm a bit confused that why Git extension is marked with `"virtualWorkspaces": true`? Maybe necessary to support remote development?

Also, if I accept extra complexity on my side, e.g., by maintaining a two-way sync-ed copy in the local tempdir, is there any chance that I could bridge with the Git extension?

Otherwise, can I put it simple that if the workspace is not backed by a local filesystem, then there's currently no way to make this workspace work with a Git extension? (Remote SSH and GitHub Repositories extension seems to be counterexample to this, but I'm not sure how they achieve such functionality)

**Reply by abrilurena** [reply]
Hi, @jjyyxx! i hope you're doing great. I'm adding @alfredodeza he might know something about this ✨

**Comment by abrilurena** [comment]
Hi, @jjyyxx! I hope you're doing great. It's been a while, so I was wondering - do you still need help with this?

**Reply by jjyyxx** [reply]
Thank you for your attention. While I'm currently not working towards this, my questions in https://github.com/microsoft/vscode-discussions/discussions/1012#discussioncomment-8321399 and https://github.com/microsoft/vscode-discussions/discussions/1012#discussioncomment-8296694 are not clarified yet, so I just leave this discussion open. At least, you could leave a definitive negative answer like `won't fix` or `not planned` ...

**Reply by abrilurena** [reply]
Hi @jjyyxx! I hope you're doing great and I'm so sorry to hear about it - i just added more people that can support you on these questions.

**Comment by robconery** [comment]
Hi @jjyyxx - I pinged the team today and they gave me a definitive WONTFIX on this based on how they work with Git. If you're up for requesting an enhancement, you can do that here: https://github.com/microsoft/vscode/issues/new/choose, just choose Feature Enhancement.

Going to close this as the question has been answered, but please feel free to re-open if I have jumped the gun.
